Sources_v3/WebSocketClient/EventMiddlewares/ChannelReadUpdaterMiddleware_Tests.swift

```diff
@@ -0,0 +1,141 @@
+        }
+        
+        // Load the channel from the db and check the if fields are correct
+        var loadedChannel: ChannelModel<DefaultDataTypes>? {
+            database.viewContext.loadChannel(cid: channelId)
+        }
+        
+        XCTAssertEqual(loadedChannel?.reads.first?.unreadMessagesCount, 10)
+        XCTAssertEqual(loadedChannel?.reads.first?.lastReadAt, Date(timeIntervalSince1970: 1))
+        
+        let newReadDate = Date(timeIntervalSince1970: 2)
+        let eventPayload = EventPayload<DefaultDataTypes>(
+            eventType: .messageRead,
+            cid: channelId,
+            user: dummyCurrentUser,
+            unreadCount: .noUnread,
+            createdAt: newReadDate
+        )
+        let messageReadEvent = try MessageReadEvent<DefaultDataTypes>(from: eventPayload)
+        
+        // Let the middleware handle the event
+        let handledEvent = try await { middleware.handle(event: messageReadEvent, completion: $0) }
+        
+        XCTAssertEqual(handledEvent?.asEquatable, messageReadEvent.asEquatable)
+        XCTAssertEqual(loadedChannel?.reads.first?.unreadMessagesCount, 0)
+        XCTAssertEqual(loadedChannel?.reads.first?.lastReadAt, newReadDate)
+    }
+    
+    func test_notificationMarkReadEvent_handledCorrectly() throws {
+        // Save a channel with a channel read
+        let channelId = ChannelId.unique
+        let payload = dummyPayload(with: channelId)
+        
+        assert(payload.channelReads.count == 1)
+        
+        try database.writeSynchronously { (session) in
+            try session.saveChannel(payload: payload)
+        }
+
+        // Load the channel from the db and check the if fields are correct
+        var loadedChannel: ChannelModel<DefaultDataTypes>? {
+            database.viewContext.loadChannel(cid: channelId)
+        }
+        
+        XCTAssertEqual(loadedChannel?.reads.first?.unreadMessagesCount, 10)
+        XCTAssertEqual(loadedChannel?.reads.first?.lastReadAt, Date(timeIntervalSince1970: 1))
+        
+        let newReadDate = Date(timeIntervalSince1970: 2)
+        let channelDetailPayload = ChannelDetailPayload<DefaultDataTypes>(
+            cid: channelId,
+            extraData: lukeExtraData,
+            typeRawValue: "",
+            lastMessageAt: nil,
+            createdAt: .unique,
+            deletedAt: nil,
+            updatedAt: .unique,
+            createdBy: nil,
+            config: .init(),
+            isFrozen: false,
+            memberCount: 0,
+            team: "",
+            members: nil
+        )
+        let eventPayload = EventPayload<DefaultDataTypes>(
+            eventType: .notificationMarkRead,
+            user: dummyCurrentUser,
+            channel: channelDetailPayload,
+            unreadCount: .noUnread,
+            createdAt: newReadDate
+        )
+        let notificationMarkReadEvent = try NotificationMarkReadEvent<DefaultDataTypes>(from: eventPayload)
```

Comment:
Again, I'd appreciate a summary comment so I don't have to try to understand all these lines.

Reply:
Added in b65ae9e402eac93dfee89c9c09afa10935dd4c21